fairlearn/metrics/_metric_frame.py

```diff
@@ -0,0 +1,515 @@
+                max_val = self.by_group[m].max()
+                result[m] = max_val
+        else:
+            result = self.by_group.groupby(level=self.control_levels).max()
+        return result
+
+    def group_min(self) -> Union[pd.Series, pd.DataFrame]:
+        """Return the minimum value of the metric over the sensitive features.
+
+        This method computes the minimum value over all combinations of
+        sensitive features for each underlying metric function in the ``by_groups``
+        property (it will only succeed if all the underlying metric
+        functions return scalar values). If there are no control
+        features, then the result will be a :class:`pandas.Series`,
+        with rows corresponding to the underying metrics.
+
+        Returns
+        -------
+        pandas.Series or pandas.DataFrame
+            If there are control features, then the result is a
+            :class:`pandas.DataFrame` with columns corresponding to the
+            underlying metrics, and rows indexed by the classes identified
+            by the control features. Otherwise, a :class:`pandas.Series`
+            indexed by the names of the underlying metrics.
+        """
+        if not self.control_levels:
+            result = pd.Series(index=self.by_group.columns, dtype='object')
+            for m in result.index:
+                min_val = self.by_group[m].min()
+                result[m] = min_val
+        else:
+            result = self.by_group.groupby(level=self.control_levels).min()
+        return result
+
+    def difference(self,
+                   method: str) -> Union[pd.Series, pd.DataFrame]:
+        """Return the maximum absolute difference between groups for each metric.
+
+        This method calculates a scalar value for each underlying metric by
+        finding the maximum absolute difference between the entries in each
+        column of the ``by_groups`` DataFrame.
+
+        Similar to other methods, the result will be a :class:`pandas.Series`
+        if there are no control features, with rows corresponding
+        to each underlying metric. If there are control features,
+        then the result will be a :class:`pandas.DataFrame` with a column
+        for each underlying metric, and rows corresponding to the subgroups
+        identified by the control feature(s).
+
+        There are two allowed values for the ``method=`` parameter. The
+        value ``between_groups`` computes the maximum difference between
+        any two pairs of groups in the ``by_groups`` DataFrame (i.e.
+        ``group_max() - group_min()``). Alternatively, ``to_overall``
+        computes the difference between each subgroup and the
+        corresponding value from ``overall`` (if there are control
+        features, then ``overall`` is multivalued for each metric).
+        The result is the absolute maximum of these values.
+
+        Returns
+        -------
+        pandas.Series or pandas.DataFrame
```

Comment:
what are the return types if only one metric is provided? i'm assuming that for one metric we return `pandas.Series` if there are some control features, and a scalar if there are no control features?

Reply:
The return types are exactly the same as if two or more metrics are provided - a Series or a DataFrame

Reply:
Also, having `.group_max()` give a scalar if there is only one result would be inconsistent with plain pandas:
```python
>>> import pandas as pd
>>> df = pd.DataFrame(data=[0,1,2,3,4])
>>> df.max()
0    4
dtype: int64
>>> type(df.max())
<class 'pandas.core.series.Series'>
```
I'm not a fan of return types changing with input data; the reason I tolerate it to some extent here is that for the no-control-features case, I agree that there's no good name for the single row of the DataFrame. However, contracting to a scalar for things like `.max()` when pandas itself does not seems unwise. And given that, I think that leaving `.overall` as a single element Series is better for consistency that quietly dropping it to a scalar.

Reply:
If `by_group` is a `pandas.Series` (as I suggest for this case) then `by_group.max()` is a scalar.

Reply:
For `by_group`, I think that we should have the following behavior:

* if `metric` is a callable then `by_group` is `pandas.Series` without any name (i.e., `by_group.name == None`)
* if `metric` is a dictionary then `by_group` is `pandas.DataFrame` (even if the dictionary has only one element)

I think that this would nicely handle both the use case when the data scientist wants to analyze a single metric as well as when `MetricFrame` is used as part of another system. It is also consistent with the pandas behavior:
```python
scores = pd.Series(data=[0, 1, 1], index=['a','b','a'])

scores.groupby(level=0).agg('mean').__class__   # pandas.Series
scores.groupby(level=0).agg(['mean','min']).__class__   # pandas.DataFrame
scores.groupby(level=0).agg(['mean']).__class__   # pandas.DataFrame
```

Reply:
Resolved to accept this proposal.